include/deal.II/fe/fe_coupling_values.h

```diff
@@ -406,6 +407,610 @@ namespace FEValuesViews
+ *
+ * FullMatrix<double> local_matrix(fe_values1.dofs_per_cell,
+ *                                 fe_values2.dofs_per_cell);
+ *
+ * // Extractor on first cell
+ * const auto v1 = cfv.first(FEValuesExtractor::Scalar(0));
+ * const auto p1 = cfv.first(FEValuesExtractor::Scalar(1));
+ *
+ * // Extractor on second cell
+ * const auto u2 = cfv.second(FEValuesExtractor::Scalar(0));
+ * const auto q2 = cfv.second(FEValuesExtractor::Scalar(1));
+ *
+ * ...
+ *
+ * for (const unsigned int q : cfv.quadrature_point_indices()) {
+ *   const auto &[x_q,y_q] = cfv.quadrature_point(q);
+ *
+ *   for (const unsigned int i : cfv.first_dof_indices())
+ *     {
+ *       const auto &v_i = cfv[v1].value(i, q);
+ *       const auto &p_i = cfv[p1].value(i, q);
+ *
+ *       for (const unsigned int j : cfv.second_dof_indices())
+ *         {
+ *           const auto &u_j = cfv[u2].value(j, q);
+ *           const auto &q_j = cfv[q2].value(j, q);
+ *
+ *           local_matrix(i, j) += (K1(x_q, y_q) * v_i * u_j +
+ *                                  K2(x_q, y_q) * p_i * q_j) *
+ *                                 cfv[0].JxW(q) *
+ *                                 cfv[1].JxW(q);
+ *         }
+ *     }
+ * }
+ * @endcode
+ *
+ * In the above loop, the quadrature points of the two FEValuesBase objects are
+ * grouped together in a single loop, while the dof indices are kept separate.
+ *
+ * Internally, this class provides an abstraction to organize coupling terms
+ * between two arbitrary FEValuesBase objects, and provides a unified way to
+ * access the values of the two basis, and of the two sets of quadrature points
+ * and weights.
+ *
+ * According to the DoFCouplingType and QuadratureCouplingType argument passed
+ * to the constructor, the class behaves differently w.r.t. how the dof
+ * indices and the quadrature points of the two different FEValuesBase objects
+ * are grouped together.
+ *
+ * The class is intended to be a higher level abstraction compared to assembling
+ * coupling terms manually.
+ *
+ * This class gives access to two different FEValuesBase objects, and it
+ * provides a unified way to access the values of the two basis, using the
+ * concept of FEValuesExtractors. Unlike the FEValuesBase class, though, we need
+ * to specify which FEValuesBase object we are referring to, and this is done by
+ * calling the first() or second() functions, which are used to *inform* the
+ * extractors about what FEValuesBase object the extractor refers to.
+ *
+ * This is achieved, in user codes, by the following snippet:
+ *
+ * @code
+ * // extract the first scalar component of the basis
+ * FEValuesExtractor scalar(0);
+ * ...
+ *
+ * FECouplingValues<dim> fe_coupling(fev1, fev1);
+ *
+ * // Extractors for the two FEValuesBase objects are returned by the first()
+ * and
+ * // second() methods
+ *
+ * const auto first = fe_coupling.first(scalar);
+ * const auto second = fe_coupling.second(scalar);
```

Comment:
Have you thought whether it's worth making the syntax something like this?
```
  fe_coupling.first()[scalar];
```
to match the usual extractor syntax? At the same time, I see that the result here is not a view as usual, but another extractor of some sort. Perhaps rename the functions so that it would read as
```
  fe_values.get_extractor_for_first(scalar);
```
or some such? (My groping for the right way to do this suggests that perhaps the function name right now is not descriptive enough for what the function actually does.)